src/ert/_c_wrappers/enkf/lark_parser.py

```diff
@@ -153,333 +185,89 @@ def substitute(
+    check_required(schema, config_dict, filename=config_file)
+
+    return config_dict
 
-    def keyword(self, tree: Tree):
-        arguments = []
-        kw = tree.children[0]
-        if not isinstance(kw, FileContextToken):
-            raise ConfigValidationError(
-                f"Unrecognized keyword {kw!r}", config_file=self.config_file
-            )
-        do_env = True
-        if kw in self.schema:
-            do_env = self.schema[kw].expand_envvar
-        instruction: Union[Instruction, JobInstruction]
-        if len(tree.children) == 2 and tree.children[1].data == "arglist":
-            node = tree.children[1]
-            name = node.children[0]
-            args = []
-            kw_list = node.children[1]
-            for kw_pair in kw_list.children:
-                if kw_pair is None:
-                    break
-                key = kw_pair.children[0]
-                val = kw_pair.children[1].children[0]
-                if not isinstance(val, FileContextToken):
-                    raise ConfigValidationError(
-                        f"Could not read keyword value {kw!r} for {key!r}",
-                        config_file=self.config_file,
-                    )
-                args.append((key, val))
-            job_name = name
-            arguments = args
-            instruction = JobInstruction(kw, job_name, arguments)
-        else:
-            for node in tree.children[1:]:
-                if node.data != "arg":
-                    raise ConfigValidationError(
-                        "Cannot mix argument list with"
-                        f" parenthesis and without in {node}"
-                    )
-
-                val = node.children[0]
-
-                if not isinstance(val, FileContextToken):
-                    raise ConfigValidationError(
-                        f"Could not read argument {val!r}",
-                        config_file=self.config_file,
-                    )
-                arguments.append(substitute(self.defines, val, expand_env=do_env))
-            instruction = Instruction(kw, arguments)
-        if kw in self.schema:
-            item = self.schema[kw]
-            if (
-                item.argc_min != CONFIG_DEFAULT_ARG_MIN
-                and len(instruction.args) < item.argc_min
-            ):
-                raise ConfigValidationError(
-                    f"{instruction.keyword!r} needs at least {item.argc_min} arguments",
-                    config_file=kw.filename,
-                )
-            if (
-                item.argc_max != CONFIG_DEFAULT_ARG_MAX
-                and len(instruction.args) > item.argc_max
-            ):
-                raise ConfigValidationError(
-                    f"{kw!r} takes maximum {item.argc_max} arguments",
-                    config_file=kw.filename,
-                )
 
-        self.keywords.append(instruction)
+def _substitute_args(args, constraints, defines):
```

Comment:
I think this could be type annotated, `_substitute` always returns a `str`, so this should return the type of args (which is ..?) OR at least a list of `str`s

Reply:
It's a bit complicated again, the actual type is dependent on the logic of what we're doing. If you are supposed to substitute, you can expect the particular item to be a FileContextToken, otherwise it can be anything. And it will return the same as given in the case of substitution, but not the other functions.

Reply:
I have added a few isinstance(x, FileContextToken) here and there, to make it clear that the function will only do stuff to FileContextToken, and otherwise return what ever was there. This is strictly not needed, if the code is correct otherwise, but does make it clearer.

Did similar another place.